include/mumble/Base64.hpp

```diff
@@ -3,30 +3,134 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
+// Inspired by https://renenyffenegger.ch/notes/development/Base64/Encoding-and-decoding-base-64-with-cpp.
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, std::numeric_limits< uint8_t >::max() + 1 > table = {
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55,
+			56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12,
+			13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64, 64, 26, 27, 28, 29, 30, 31, 32,
+			33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
+
+		if (in.empty() || in.size() % 4) {
+			return 0;
+		}
+
+		size_t size = in.size() / 4 * 3;
```

Comment:
```suggestion
		size_t size = (in.size() / 4) * 3;
```
This makes precedence clearer